src/client/sandbox/cookie/window-sync.ts

```diff
@@ -1,113 +1,173 @@
         try {
             // eslint-disable-next-line no-restricted-properties
-            return win[INTERNAL_PROPS.hammerhead].sandbox.cookie;
+            const cookieSandbox = win[INTERNAL_PROPS.hammerhead].sandbox.cookie;
+
+            return cookieSandbox.document && cookieSandbox;
         }
         catch (e) {
             return null;
         }
     }
 
-    _getWindowsForSync (initiator) {
-        const windows = [];
+    private static _removeAllWindowSyncCookie () {
+        const cookies       = nativeMethods.documentCookieGetter.call(document);
+        const parsedCookies = parseClientSyncCookieStr(cookies);
+        const sessionId     = settings.get().sessionId;
 
-        if (this.win !== this.win.parent && this.win.parent !== initiator)
-            windows.push(this.win.parent);
+        for (const outdatedCookie of parsedCookies.outdated)
+            nativeMethods.documentCookieSetter.call(document, generateDeleteSyncCookieStr(outdatedCookie));
 
-        for (const frameWin of this.win.window) {
-            if (frameWin !== initiator)
-                windows.push(frameWin);
-        }
+        for (const parsedCookie of parsedCookies.actual) {
+            if (parsedCookie.sid === sessionId && parsedCookie.isWindowSync) {
+                nativeMethods.documentCookieSetter.call(document, generateDeleteSyncCookieStr(parsedCookie));
 
-        return windows;
+                if (parsedCookie.isClientSync) {
+                    changeSyncType(parsedCookie, { window: false });
+                    nativeMethods.documentCookieSetter.call(document, formatSyncCookie(parsedCookie));
+                }
+            }
+        }
     }
 
-    syncBetweenWindows (cookies, initiator, callback) {
-        if (!cookies.length) {
-            if (callback)
-                callback();
+    private _onMsgReceived ({ message, source }) {
+        if (message.cmd === SYNC_COOKIE_START_CMD) {
+            this._cookieSandbox.syncWindowCookie(message.cookies);
 
-            return null;
+            if (this._win !== this._win.top)
+                this._messageSandbox.sendServiceMsg({ id: message.id, cmd: SYNC_COOKIE_DONE_CMD }, source);
+            else
+                this.syncBetweenWindows(message.cookies, source);
+        }
+        else if (message.cmd === SYNC_COOKIE_DONE_CMD) {
+            if (this._resolversMap.has(message.id))
+                this._resolversMap.get(message.id)();
         }
+    }
 
-        const windowsForSync = this._getWindowsForSync(initiator);
-        const syncMessages   = [];
+    private _getWindowsForSync (initiator: Window, currentWindow: Window = this._win.top, windows: Array<Window> = []) {
+        if (currentWindow !== initiator && currentWindow !== this._win.top)
+            windows.push(currentWindow);
 
-        for (const win of windowsForSync) {
-            const cookieSandbox = WindowSync._getCookieSandbox(win);
+        for (let i = 0; i < currentWindow.frames.length; i++)
```

Comment:
Can we use `for ..of` loop?

Reply:
`error TS2488: Type 'Window' must have a '[Symbol.iterator]()' method that returns an iterator.`
But we have the plugin for `for .. of` statement. I can use `for .. of` with the `@ts-ignore` instruction.